### PR TITLE
Inventory on string metrics

### PR DESCRIPTION
**Pull request by jotak**
(Got all integration tests green on my machine, now checking with travis...)

### CODE REVIEW

**Comment by jotak** [comment]
... need to set org.hawkular.metrics.string-size to -1 in metrics config db...

**Comment by jmazzitelli** [comment]
Putting a hold on any changes/PR merges to the agent until we decide what to do about this major refactoring effort going on right now (see PR #302 ). I've got over 200 files changed (many are wholesale moving of large chunks of files) and any thing going in now it going to potentially involve major conflicts. To avoid any git pain, going to wait until that major refactor is complete. Should be soon.

**Comment by jotak** [comment]
Ok, anyway, this change must also be coordinated with the MIQ side and other parts (hServices...) so it's not ready to be merged so fast

**Comment by jmazzitelli** [comment]
The agent refactor is complete and merged into master. As expected, this PR is affected. Major conflict resolution required.

File: hawkular-agent-core/src/main/java/org/hawkular/agent/monitor/service/AgentCoreEngine.java [hunk 1]
```diff
@@ -291,15 +288,6 @@ public void startHawkularAgent(AgentCoreEngineConfiguration newConfiguration) {
             switch (this.configuration.getStorageAdapter().getType()) {
                 case HAWKULAR:
                     // if we are participating in a full Hawkular environment, we need to do some additional things:
-                    // 1. register our feed ID
-                    // 2. connect to the server's feed comm channel
-                    try {
-                        registerFeed(tenantIds);
```
Comment (jotak) [inline hawkular-agent-core/src/main/java/org/hawkular/agent/monitor/service/AgentCoreEngine.java:297]: @jmazzitelli : I removed registerFeed as it doesn't have to be explicitly setup now. Except if you think that a feed should have an existence even if it has nothing associated with (no metric type, no resource type, no root resource)... in which case I should rollback this part.

File: hawkular-agent-core/src/main/java/org/hawkular/agent/monitor/storage/AsyncInventoryStorage.java [hunk 2]
```diff
@@ -708,31 +704,20 @@ private void postInventoryData(InventoryMetric metric, String jsonPayload, Strin
...
-                "\"dataRetention\": 90," +
-                "\"tags\": {" +
-                "\"module\": \"inventory\"," +
-                "\"feed\": \"" + metric.getFeed() + "\"," +
-                "\"type\": \"" + metric.getType() + "\"";
+    private void setupMetric(String tenant, InventoryMetric metric, Map<String, Collection<String>> resourceTypes) throws Exception {
+        MetricDefinition def = metric.toMetricDefinition();
         if (resourceTypes != null) {
             for (Map.Entry<String, Collection<String>> entry : resourceTypes.entrySet()) {
                 // FIXME: make sure "," cannot be used in ids, or choose another character
```
Comment (jotak) [inline hawkular-agent-core/src/main/java/org/hawkular/agent/monitor/storage/AsyncInventoryStorage.java]: (note for self: need a fix here)

**Comment by jotak** [comment]
@jmazzitelli can you have a look at the last commit? ( https://github.com/hawkular/hawkular-agent/pull/303/commits/6000ce76bb66376445513b63f8fc43b166fd5adc )

Following our discussion yesterday, I made some changes to avoid unnecessary updates. Had to refactor some code.

File: hawkular-agent-core/src/main/java/org/hawkular/agent/monitor/api/InventoryEvent.java [hunk 3]
```diff
@@ -16,58 +16,156 @@
...
+    private final Map<String, Resource<L>> addedOrModifiedRootResources;
+    private final Map<String, Resource<L>> removedRootResources;
+    private final List<Resource<L>> addedOrModified;
+    private final List<Resource<L>> removed;
 
     /**
-     * Creates an inventory event.
-     *
-     * @param samplingService a service that provides details such as feed ID and endpoint information that helps
+     * Creates a discovery event.
```
Comment (jotak) [inline hawkular-agent-core/src/main/java/org/hawkular/agent/monitor/api/InventoryEvent.java]: Will update javadoc

File: hawkular-agent-core/src/main/java/org/hawkular/agent/monitor/api/InventoryEvent.java [hunk 4]
```diff
@@ -16,58 +16,156 @@
...
+
+    /**
+     * @return the map of added or modified resources during this discovery
+     */
+    public Collection<Resource<L>> getAddedOrModifiedRootResources() {
+        return addedOrModifiedRootResources.values();
+    }
+
+    /**
+     * @return the map of removed resources during this discovery
```
Comment (jmazzitelli) [inline hawkular-agent-core/src/main/java/org/hawkular/agent/monitor/api/InventoryEvent.java]: The javadoc for this and the one above (getRemovedRootResources, getAddedOrModifiedRootResources) seems wrong. These return the root resources of all the resources that were removed (or added/modified). These root resources weren't necessarily added or removed (though they could have been, too), but they are the root resources of the resources that were added or removed. Right?

File: hawkular-agent-core/src/main/java/org/hawkular/agent/monitor/api/InventoryEvent.java [hunk 5]
```diff
@@ -16,58 +16,156 @@
...
+     */
+    public Collection<Resource<L>> getRemovedRootResources() {
+        return removedRootResources.values();
+    }
+
+    public List<Resource<L>> getAddedOrModified() {
+        return addedOrModified;
+    }
+
+    public List<Resource<L>> getRemoved() {
```
Comment (jmazzitelli) [inline hawkular-agent-core/src/main/java/org/hawkular/agent/monitor/api/InventoryEvent.java]: Here you should add javadoc to these two (getAddedOrModified, getRemoved) explaining that these are the actual resources that were added/modified and removed (which is different than the getXYZRootResources methods).

**Review by jmazzitelli — CHANGES_REQUESTED** [review]

File: hawkular-agent-core/src/main/java/org/hawkular/agent/monitor/api/InventoryEvent.java (hunk 4 above)
Comment (jotak) [reply hawkular-agent-core/src/main/java/org/hawkular/agent/monitor/api/InventoryEvent.java]: Yes, that's correct. I'm doing a javadoc cleaning pass

File: hawkular-agent-core/src/main/java/org/hawkular/agent/monitor/api/InventoryEvent.java (hunk 3 above)
Comment (jotak) [reply hawkular-agent-core/src/main/java/org/hawkular/agent/monitor/api/InventoryEvent.java]: done

File: hawkular-agent-core/src/main/java/org/hawkular/agent/monitor/api/InventoryEvent.java (hunk 4 above)
Comment (jotak) [reply hawkular-agent-core/src/main/java/org/hawkular/agent/monitor/api/InventoryEvent.java]: done

File: hawkular-agent-core/src/main/java/org/hawkular/agent/monitor/api/InventoryEvent.java [hunk 6]
```diff
@@ -16,58 +16,190 @@
...
+        this.resourceTypeManager = resourceTypeManager;
+        this.addedOrModified = addedOrModified;
+        this.removed = removed;
+
+        // Distribute 'addedOrModified' and 'removed' in 'addedOrModifiedRootResources' and 'removedRootResources'
+        addedOrModifiedRootResources = new HashMap<>();
+        removedRootResources = new HashMap<>();
+        addedOrModified.forEach(r -> {
+            Resource<L> root = getRootResource(r);
+            addedOrModifiedRootResources.put(root.getID().getIDString(), root);
```
Comment (jmazzitelli) [inline hawkular-agent-core/src/main/java/org/hawkular/agent/monitor/api/InventoryEvent.java:84]: Lines 83 and 84 tell me that "addedOrModifiedRootResources" will contain the roots of all resources that were added or modified. Is this what you intended? So if a resource changes one of its resource config properties (i.e. it is modified) then its root resource will go into this addedOrModifiedRootResources map (in that case the root isn't really modified - its list of descendants didn't change with a new resource added or old resource removed - it is just that one of its descendants (child, grandchild, or lower) got modified - in this case, do you want the root resource to be considered "modified")?

Comment (jotak) [reply hawkular-agent-core/src/main/java/org/hawkular/agent/monitor/api/InventoryEvent.java:84]: Yes, this is what is intended. The reason is that the whole root resource subtree is going to be persisted. So we need to separate root resources that have 1 modification (at least) wherever in their subtree = `addedOrModifiedRootResources` map ; and root resources that need to be entirely deleted = `removedRootResources` map.

File: hawkular-agent-core/src/main/java/org/hawkular/agent/monitor/storage/AsyncInventoryStorage.java [hunk 7]
```diff
@@ -451,188 +378,156 @@ private String getInventoryId(NamedObject no) {
...
-        Map<Resource<L>, Offline<org.hawkular.inventory.api.model.Resource.Blueprint>> rBlueprints;
-        rBlueprints = bldr.build(resourceManager);
-        rBlueprints.forEach((r, bp) -> performResourceSync(bp, tenantIdToUse, resourceManager.size(r)));
+        Set<ID> addedOrModifiedIds = event.getAddedOrModifiedRootResources().stream()
+                .map(IDObject::getID)
+                .collect(Collectors.toSet());
+        resourceManager.getRootResources().forEach(r -> {
+            // Ignore unmodified resources that doesn't need to be refreshed
+            if (addedOrModifiedIds.contains(r.getID()) || needToRefresh(r)) {
+                log.infof("Updating root resource: %s", r.getID().getIDString());
```
Comment (jotak) [inline hawkular-agent-core/src/main/java/org/hawkular/agent/monitor/storage/AsyncInventoryStorage.java]: @jmazzitelli are you fine with those 3 log.info (also line 444 and 435) that happen every time inventory storage is updated? That can be interesting for tracing http calls, but if you prefer I can put to debug level (or remove).

Comment (jmazzitelli) [reply hawkular-agent-core/src/main/java/org/hawkular/agent/monitor/storage/AsyncInventoryStorage.java]: you can move them to debug level

**Comment by jmazzitelli** [comment]
@metlos took this PR and built upon it and that is now PR #320 which supercedes this PR. This PR will be closed.

**Comment by jmazzitelli** [comment]
PR #325 is the one we are going to use - close this one

**Pull request closed without merging**